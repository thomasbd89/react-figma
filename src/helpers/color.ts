import { Color } from '../types';

export const colorToRGB = (color: Color): RGB => {
    if (color[0] === '#') {
        let colorCode = color.substr(1);

        const colorInt = parseInt(colorCode, 16);

        return {
            r: ((colorInt >> 16) & 255) / 255.0,
            g: ((colorInt >> 8) & 255) / 255.0,
            b: (colorInt & 255) / 255.0
        };
    }
};

export const colorToRGBA = (color: Color, opacity: number): RGBA => {
    return {
        ...colorToRGB(color),
        a: opacity
    };
};
