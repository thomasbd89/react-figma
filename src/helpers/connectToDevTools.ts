import { connectToDevTools as connectToDevToolsInternal } from 'react-devtools-core';

export const connectToDevTools = () => {
    // Try construction is used to avoid crashing if devtools is not launched
    try {
        connectToDevToolsInternal({
            isAppActive: () => true,
            host: 'localhost',
            port: 8097,
            resolveRNStyle: null
        });
    } catch (e) {}
};
