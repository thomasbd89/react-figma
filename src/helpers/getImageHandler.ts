import { messagePromise } from './messagePromise';

export const getImageHandler = async (data: string): Promise<Image> => {
    const response = await messagePromise({ type: 'base64ToBlobWorker', value: data });
    return figma.createImage(response);
};
