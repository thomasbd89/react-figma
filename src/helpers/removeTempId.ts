export const removeTempId = node => {
    if (node && node.pluginData) {
        delete node.pluginData.tempId;
    }
    if (node && node.children) {
        node.children.forEach(removeTempId);
    }
    return node;
};
