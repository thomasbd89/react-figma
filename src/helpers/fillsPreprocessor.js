var __awaiter =
    (this && this.__awaiter) ||
    function(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function(resolve, reject) {
            function fulfilled(value) {
                try {
                    step(generator.next(value));
                } catch (e) {
                    reject(e);
                }
            }
            function rejected(value) {
                try {
                    step(generator['throw'](value));
                } catch (e) {
                    reject(e);
                }
            }
            function step(result) {
                result.done
                    ? resolve(result.value)
                    : new P(function(resolve) {
                          resolve(result.value);
                      }).then(fulfilled, rejected);
            }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    };
import { getImageHandler } from './getImageHandler';
export const fillsPreprocessor = props =>
    __awaiter(this, void 0, void 0, function*() {
        if (props.fills) {
            const newFills = props.fills.map(fill =>
                __awaiter(this, void 0, void 0, function*() {
                    if (fill.type !== 'IMAGE' || !fill.image) {
                        return fill;
                    }
                    const mappedFill = Object.assign({}, fill);
                    const imageHandler = yield getImageHandler(mappedFill.image);
                    delete mappedFill.image;
                    mappedFill.imageHash = imageHandler.hash;
                    mappedFill.scaleMode = mappedFill.scaleMode || 'FILL';
                    return mappedFill;
                })
            );
            return yield Promise.all(newFills);
        }
        return [];
    });
