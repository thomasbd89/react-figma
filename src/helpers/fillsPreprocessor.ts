import { getImageHandler } from './getImageHandler';

export const fillsPreprocessor = async (props): Promise<Array<Paint>> => {
    if (props.fills) {
        const newFills = props.fills.map(async fill => {
            if (fill.type !== 'IMAGE' || !fill.image) {
                return fill;
            }

            const mappedFill = { ...fill };

            const imageHandler = await getImageHandler(mappedFill.image);
            delete mappedFill.image;

            mappedFill.imageHash = imageHandler.hash;
            mappedFill.scaleMode = mappedFill.scaleMode || 'FILL';

            return mappedFill;
        });

        return await Promise.all(newFills);
    }

    return [];
};
