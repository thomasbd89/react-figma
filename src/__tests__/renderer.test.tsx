import * as React from 'react';
import { render } from '../renderer';
import { Rectangle, Page, Text, Group, Frame, Svg, createComponent } from '..';
import { createFigma } from 'figma-api-stub';
import { Subject } from 'rxjs';
import { delay, take } from 'rxjs/operators';
import { wait } from '../helpers/wait';
import '../rpc';
import { removeTempId } from '../helpers/removeTempId';
import { removeNodeBatchId } from '../helpers/removeNodeBatchId';

const removeMeta = node => {
    return removeNodeBatchId(removeTempId(node));
};

describe('renderer', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('render single component', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
            </Page>
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(1);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('rerender single component', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
            </Page>
        );
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(1);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('insert new component between', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#005aff' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(3);
        expect(figma.createPage).toHaveBeenCalledTimes(1);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('remove component between', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                <Text style={{ width: 200, height: 100 }} characters="test" />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(2);
        expect(figma.createText).toHaveBeenCalledTimes(1);
        expect(figma.createPage).toHaveBeenCalledTimes(1);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('remove component between (equal components)', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#0048ff' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#00ff00' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#0048ff' }} />
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(3);
        expect(figma.createPage).toHaveBeenCalledTimes(1);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('text instance without Text component', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} /> fff
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(2);
        expect(figma.createText).toHaveBeenCalledTimes(0);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('text instance without Text component (with hydration page)', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        await render(<Page></Page>);
        await render(
            <Page>
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} /> fff
                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff3500' }} />
            </Page>
        );
        expect(figma.createRectangle).toHaveBeenCalledTimes(2);
        expect(figma.createText).toHaveBeenCalledTimes(0);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('creates single group', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        figma.group = jest.fn().mockImplementation(figma.group);

        await render(
            <Page>
                <Group>
                    <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                </Group>
            </Page>
        );

        expect(figma.group).toHaveBeenCalledTimes(1);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('creates nested groups', async () => {
        figma.createRectangle = jest.fn().mockImplementation(figma.createRectangle);
        figma.createText = jest.fn().mockImplementation(figma.createText);
        figma.createPage = jest.fn().mockImplementation(figma.createPage);
        figma.group = jest.fn().mockImplementation(figma.group);

        await render(
            <Page>
                <Group>
                    <Group>
                        <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                    </Group>
                    <Group>
                        <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                        <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                    </Group>
                </Group>
            </Page>
        );

        expect(figma.group).toHaveBeenCalledTimes(3);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('Groups inserting', async () => {
        const waiting = new Subject();

        const Component = () => {
            const [flag, setFlag] = React.useState(false);
            React.useEffect(() => {
                setTimeout(() => {
                    setFlag(true);
                    waiting.next();
                });
            }, []);

            return (
                <Page>
                    <Frame>
                        <Group>
                            <Rectangle style={{ width: 200, height: 100, backgroundColor: '#0054ff' }} />
                        </Group>
                        {flag && (
                            <Group>
                                <Rectangle style={{ width: 200, height: 100, backgroundColor: '#ff0017' }} />
                            </Group>
                        )}
                        <Group>
                            <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                        </Group>
                    </Frame>
                </Page>
            );
        };

        await render(<Component />);

        return new Promise(resolve => {
            waiting
                .pipe(
                    take(1),
                    delay(0)
                )
                .subscribe(() => {
                    expect(removeMeta(figma.root)).toMatchSnapshot();
                    resolve();
                });
        });
    });

    it('mark page isCurrent=true', () => {
        render(<Page isCurrent name={'New page'} />);
        expect(figma.currentPage).toMatchSnapshot();
    });

    it('mark page isCurrent=false', () => {
        render(<Page name={'New page'} />);
        expect(figma.currentPage).toMatchSnapshot();
    });

    it('Text component supported text instance children', async () => {
        figma.createText = jest.fn().mockImplementation(figma.createText);
        await render(
            <Page>
                <Text>Some text</Text>
            </Page>
        );
        await wait();
        expect(figma.createText).toHaveBeenCalledTimes(1);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('Text instance hydration', async () => {
        figma.createText = jest.fn().mockImplementation(figma.createText);
        await render(
            <Page>
                <Text>Some text</Text>
            </Page>
        );
        await wait();
        await render(
            <Page>
                <Text>Some text 2</Text>
            </Page>
        );
        await wait();

        expect(figma.createText).toHaveBeenCalledTimes(1);
        expect(removeMeta(figma.root)).toMatchSnapshot();
    });

    it('Text characters applied', async () => {
        render(<Text characters="some text" />);
        await wait();
        await wait();
        expect(figma.root).toMatchSnapshot();
    });

    it('Text with custom font applied', async () => {
        render(<Text style={{ fontFamily: 'Helvetica', fontWeight: 'bold' }}>some text</Text>);

        await wait();
        await wait();
        expect(figma.root).toMatchSnapshot();
    });

    it('Text instance updating', async () => {
        const waiting = new Subject();
        figma.createText = jest.fn().mockImplementation(figma.createText);

        const Component = () => {
            const [text, setText] = React.useState('text 1');
            React.useEffect(() => {
                setTimeout(() => {
                    setText('text 2');
                    waiting.next();
                });
            }, []);

            return <Text>{text}</Text>;
        };

        render(<Component />);

        return new Promise(resolve => {
            waiting.pipe(take(1)).subscribe(() => {
                expect(figma.root).toMatchSnapshot();
                resolve();
            });
        });
    });

    it('Svg render', () => {
        figma.createNodeFromSvg = jest.fn(source => {
            const rect = figma.createRectangle();
            rect.fills = [
                {
                    color: {
                        b: 1,
                        g: 0,
                        r: 0
                    },
                    type: 'SOLID'
                }
            ];

            const frame = figma.createFrame();
            frame.appendChild(rect);
            return frame;
        });

        render(<Svg source={'<svg />'} />);
        expect(figma.createNodeFromSvg).toHaveBeenCalledTimes(1);
        expect(figma.root).toMatchSnapshot();
    });

    it('Svg hydration', () => {
        figma.createNodeFromSvg = jest.fn(source => {
            const rect = figma.createRectangle();
            rect.fills = [
                {
                    color: {
                        b: source === 'source1' ? 1 : 0,
                        g: source === 'source2' ? 1 : 0,
                        r: 0
                    },
                    type: 'SOLID'
                }
            ];

            const frame = figma.createFrame();
            frame.appendChild(rect);
            return frame;
        });

        render(<Svg source="source1" />);
        render(<Svg source="source2" />);
        expect(figma.createNodeFromSvg).toHaveBeenCalledTimes(2);
        expect(figma.root).toMatchSnapshot();
    });

    it('Svg instance updating', async () => {
        figma.createNodeFromSvg = jest.fn(source => {
            const rect = figma.createRectangle();
            rect.fills = [
                {
                    color: {
                        b: source === 'source1' ? 1 : 0,
                        g: source === 'source2' ? 1 : 0,
                        r: 0
                    },
                    type: 'SOLID'
                }
            ];

            const frame = figma.createFrame();
            frame.appendChild(rect);
            return frame;
        });

        const waiting = new Subject();
        const Component = () => {
            const [source, setSource] = React.useState('source1');
            React.useEffect(() => {
                setTimeout(() => {
                    setSource('source2');
                    waiting.next();
                });
            }, []);

            return <Svg source={source} />;
        };

        render(<Component />);

        return new Promise(resolve => {
            waiting.pipe(take(1)).subscribe(() => {
                expect(figma.createNodeFromSvg).toHaveBeenCalledTimes(2);
                expect(figma.root).toMatchSnapshot();
                resolve();
            });
        });
    });

    it('createComponent basic', () => {
        const Rect = createComponent();
        render(
            <>
                <Rect.Component>
                    <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                </Rect.Component>
                <Rect.Instance />
            </>
        );
        expect(figma.root).toMatchSnapshot();
    });

    it('createComponent hydration', () => {
        const Rect = createComponent();
        render(
            <>
                <Rect.Component>
                    <Rectangle style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                </Rect.Component>
                <Rect.Instance />
            </>
        );
        render(
            <>
                <Rect.Component>
                    <Rectangle style={{ width: 200, height: 100, backgroundColor: '#0051ff' }} />
                </Rect.Component>
                <Rect.Instance />
            </>
        );
        expect(figma.root).toMatchSnapshot();
    });

    it('createComponent overriding', async () => {
        const Rect = createComponent();
        render(
            <>
                <Rect.Component>
                    <Rectangle name="rect" style={{ width: 200, height: 100, backgroundColor: '#12ff00' }} />
                </Rect.Component>
                <Rect.Instance
                    overrides={{
                        rect: {
                            style: {
                                backgroundColor: '#0000ff'
                            }
                        }
                    }}
                />
            </>
        );
        await wait();
        expect(figma.root).toMatchSnapshot();
    });
});
