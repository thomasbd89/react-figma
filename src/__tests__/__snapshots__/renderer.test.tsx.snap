// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer Text component supported text instance children 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "characters": "Some text",
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{}",
          },
          "type": "TEXT",
        },
      ],
      "parent": [Circular],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer Text instance hydration 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "characters": "Some text 2",
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{}",
          },
          "type": "TEXT",
        },
      ],
      "parent": [Circular],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer Text instance updating 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "characters": "text 2",
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{}",
          },
          "type": "TEXT",
        },
      ],
      "parent": [Circular],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer creates nested groups 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        ChildrenMixinStub {
          "children": Array [
            ChildrenMixinStub {
              "children": Array [
                LayoutMixinStub {
                  "fills": Array [
                    Object {
                      "color": Object {
                        "b": 0,
                        "g": 1,
                        "r": 0.07058823529411765,
                      },
                      "type": "SOLID",
                    },
                  ],
                  "height": 100,
                  "parent": [Circular],
                  "pluginData": Object {
                    "isReactFigmaNode": "true",
                    "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
                  },
                  "type": "RECTANGLE",
                  "width": 200,
                },
              ],
              "parent": [Circular],
              "pluginData": Object {
                "isReactFigmaNode": "true",
                "reactStyle": "{}",
              },
              "type": "GROUP",
            },
            ChildrenMixinStub {
              "children": Array [
                LayoutMixinStub {
                  "fills": Array [
                    Object {
                      "color": Object {
                        "b": 0,
                        "g": 1,
                        "r": 0.07058823529411765,
                      },
                      "type": "SOLID",
                    },
                  ],
                  "height": 100,
                  "parent": [Circular],
                  "pluginData": Object {
                    "isReactFigmaNode": "true",
                    "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
                  },
                  "type": "RECTANGLE",
                  "width": 200,
                },
                LayoutMixinStub {
                  "fills": Array [
                    Object {
                      "color": Object {
                        "b": 0,
                        "g": 1,
                        "r": 0.07058823529411765,
                      },
                      "type": "SOLID",
                    },
                  ],
                  "height": 100,
                  "parent": [Circular],
                  "pluginData": Object {
                    "isReactFigmaNode": "true",
                    "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
                  },
                  "type": "RECTANGLE",
                  "width": 200,
                },
              ],
              "parent": [Circular],
              "pluginData": Object {
                "isReactFigmaNode": "true",
                "reactStyle": "{}",
              },
              "type": "GROUP",
            },
          ],
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{}",
          },
          "type": "GROUP",
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer creates single group 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        ChildrenMixinStub {
          "children": Array [
            LayoutMixinStub {
              "fills": Array [
                Object {
                  "color": Object {
                    "b": 0,
                    "g": 1,
                    "r": 0.07058823529411765,
                  },
                  "type": "SOLID",
                },
              ],
              "height": 100,
              "parent": [Circular],
              "pluginData": Object {
                "isReactFigmaNode": "true",
                "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
              },
              "type": "RECTANGLE",
              "width": 200,
            },
          ],
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{}",
          },
          "type": "GROUP",
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer insert new component between 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "characters": "test",
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{}",
          },
          "type": "TEXT",
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer mark page isCurrent=false 1`] = `
ChildrenMixinStub {
  "children": Array [],
  "parent": ChildrenMixinStub {
    "children": Array [
      [Circular],
      ChildrenMixinStub {
        "children": Array [],
        "name": "New page",
        "parent": [Circular],
        "pluginData": Object {
          "isReactFigmaNode": "true",
          "reactStyle": "{}",
        },
        "type": "PAGE",
      },
    ],
    "type": "DOCUMENT",
  },
  "type": "PAGE",
}
`;

exports[`renderer mark page isCurrent=true 1`] = `
ChildrenMixinStub {
  "children": Array [],
  "name": "New page",
  "parent": ChildrenMixinStub {
    "children": Array [
      ChildrenMixinStub {
        "children": Array [],
        "parent": [Circular],
        "type": "PAGE",
      },
      [Circular],
    ],
    "type": "DOCUMENT",
  },
  "pluginData": Object {
    "isReactFigmaNode": "true",
    "reactStyle": "{}",
  },
  "type": "PAGE",
}
`;

exports[`renderer remove component between (equal components) 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 1,
                "g": 0.2823529411764706,
                "r": 0,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#0048ff\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer remove component between 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer render single component 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer rerender single component 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer text instance without Text component (with hydration page) 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`renderer text instance without Text component 1`] = `
ChildrenMixinStub {
  "children": Array [
    ChildrenMixinStub {
      "children": Array [],
      "parent": [Circular],
      "type": "PAGE",
    },
    ChildrenMixinStub {
      "children": Array [
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 1,
                "r": 0.07058823529411765,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#12ff00\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
        LayoutMixinStub {
          "fills": Array [
            Object {
              "color": Object {
                "b": 0,
                "g": 0.20784313725490197,
                "r": 1,
              },
              "type": "SOLID",
            },
          ],
          "height": 100,
          "parent": [Circular],
          "pluginData": Object {
            "isReactFigmaNode": "true",
            "reactStyle": "{\\"width\\":200,\\"height\\":100,\\"backgroundColor\\":\\"#ff3500\\"}",
          },
          "type": "RECTANGLE",
          "width": 200,
        },
      ],
      "parent": [Circular],
      "pluginData": Object {
        "isReactFigmaNode": "true",
        "reactStyle": "{}",
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;
